python/cudf/cudf/_lib/copying.pyx

```diff
@@ -735,3 +737,196 @@ def segmented_gather(Column source_column, Column gather_map):
+
+        p.c_obj = move(data_)
+        p.column_names = header["column-names"]
+        p.index_names = header["index-names"]
+
+        return p
+
+    @staticmethod
+    cdef PackedColumns c_from_py_table(Table input_table, keep_index=True):
+        """
+        Construct a PackedColumns object from a ``cudf.DataFrame``.
+        """
+        from cudf.core import RangeIndex
+
+        cdef PackedColumns p = PackedColumns.__new__(PackedColumns)
+
+        if keep_index and not input_table.index.equals(
+            RangeIndex(start=0, stop=len(input_table), step=1)
+        ):
+            input_table_view = input_table.view()
+            p.index_names = input_table._index_names
+        else:
+            input_table_view = input_table.data_view()
+
+        p.c_obj = move(cpp_copying.pack(input_table_view))
+        p.column_names = input_table._column_names
+
+        return p
+
+    cdef Table c_unpack(self):
+        return Table.from_table_view(
+            cpp_copying.unpack(self.c_obj),
+            self,
+            self.column_names,
+            self.index_names
+        )
+
+    cdef const void* c_metadata_ptr(self) except *:
+        return self.c_obj.metadata_.get()[0].data()
+
+    cdef size_t c_metadata_size(self) except *:
+        return self.c_obj.metadata_.get()[0].size()
+
+    cdef void* c_gpu_data_ptr(self) except *:
+        return self.c_obj.gpu_data.get()[0].data()
+
+    cdef size_t c_gpu_data_size(self) except *:
+        return self.c_obj.gpu_data.get()[0].size()
+
+
+def pack(Table input_table, keep_index=True):
+    """
+    Pack the columns of a ``cudf.DataFrame`` into a single GPU memory buffer.
+
+    Parameters
+    ----------
+    input_table : DataFrame
+        DataFrame to return packed representation of.
+
+    keep_index : boolean, default True
+        Pack the DataFrame's index columns along with the data columns.
+        This step will implicitly be skipped if the index columns can be
+        reconstructed on the fly at unpacking time.
+
+    Returns
+    -------
+    PackedColumns
+        A packed representation of ``input_table``.
+
+    Examples
+    --------
+    Build a dataframe and pack it:
+
+    >>> import cudf
+    >>> from cudf._lib.copying import pack
+    >>> df = cudf.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
+    >>> df
+        a   b
+    0   1   4
+    1   2   5
+    2   3   6
+    >>> packed = pack(df)
+    >>> packed
+    <cudf._lib.copying.PackedColumns at 0x7f34ff6a4b80>
+    """
+    return PackedColumns.c_from_py_table(input_table, keep_index)
+
+
+def unpack(PackedColumns packed):
+    """
+    Unpack the results of packing a ``cudf.DataFrame``, returning a new
+    ``DataFrame`` in the process.
+
+    Parameters
+    ----------
+    packed : PackedColumns
+        The results of calling ``pack()`` on a ``cudf.DataFrame``.
+
+    Returns
+    -------
+    cudf.DataFrame
+        A copy of the original DataFrame which was packed.
```

Comment:
Want to make sure: is it `Table.from_table_view` that performed the copy? Because libcudf `unpack` did not.

Reply:
Yes - if we end up keeping the detailed docstrings, is there a better way I could phrase this to emphasize that the copy operation is not a consequence of `unpack`?

Reply:
I don't think that's necessary, since whichever function makes a copy of the input object is a mere implementation detail. Thanks for answering my question though.